framework/doc/content/documentation/application_development/debugging.md

```diff
@@ -0,0 +1,164 @@
+# Debugging
+
+At some point while developing a MOOSE-based application you will probably need to use a debugger.  A debugger will allow you to stop your program at certain points (or if things happen such as a memory segfault).  Once stopped you can then carefully step through the program, inspecting variable values as you go in order to find the source of the problem.
+
+In particular, if you ever see a "Segfault" or a "Signal 11" that means it's time to pull out the debugger.  Any debugger will automatically stop once a segfault is reached, showing you exactly where the invalid memory access occurred.
+
+For a good tutorial on debugging [see Example 21](/ex21_debugging.md)
+
+## Debug Executable
+
+The first step to debugging anything is to build a debug executable.  By default MOOSE-based applications are built in "optimized" (`opt`) mode.  That ensures the fastest solves.  However, an optimized executable is missing a lot of information that is useful to a debugger and the optimization process itself can cause code to get reordered (or even skipped!) making it difficult to step through a program.
+
+To build an executable suitable for debugging you need to set the `METHOD` environment variable to `dbg`.  You can `export` it in your environment but it's usually simpler to use a UNIX shortcut that allows you to define environment variables at the same time you run a command, like so:
+
+```bash
+METHOD=dbg make -j 8
+```
+
+Always remember that the `8` should be modified to reflect the number of processors you want to use for the build (usually the number of cores in your computer).
+
+Once the build is complete you should end up with a "debug executable" that look like: `yourapp-dbg`.  That executable is perfect for loading into a debugger.  However, that executable will run VERY slowly - so make sure that before you begin debugging you come up with a problem that is as small as possible but still shows the problem you're trying to fix.
+
+## Debuggers
+
+Many differen debuggers exist: `lldb`, `gdb`, `ddd`, Totalview and Intel Debugger are just a few.  While a command-line debugger (like `lldb` or `gdb`) might seem daunting at first, they are an invaluable tool for quick debugging and debugging in complicated scenarios such as when you're running on a cluster.  Learning one should be essential to any computational scientist.
+
+For debugging MOOSE-based applications we recommend `lldb` if you're using the clang compiler (default on Mac OSX) and `gdb` for the `gcc` compiler (default on Linux).
+
+### LLDB and GDB
+
+`lldb` and `gdb` are very similar.  They work on the "command-line" taking text input and moving through your program as it executes.
+
+With our MOOSE package on Mac OSX you actually need to run lldb using `sudo` so it has the elevated privileges it needs to attach to your program.  To invoke `lldb` with a MOOSE-based application on Mac OSX you would do:
```

Comment:
I'm confused. I've never run lldb with sudo. Works fine as an unprivileged user for me.

Reply:
Two things:

1.  Yes you do.  Every time you run `lldb` (if you haven't sudoed for a little while) you get a popup message on OSX that asks you your password.  You're so used to it that you just don't remember.

2.  With the new MOOSE package we now bundle in an `lldb` that doesn't do that popup.  To use it you have to sudo explicitly.

Reply:
You use sudo without a password on your machine?!

Reply:
Well - only for this one command for parallel debugging (look at the docs here for an explanation).  There is currently no other way to do parallel debugging using the newest MOOSE package.

Reply:
On my mac the stock lldb asks only once for a password (per login session?).

Reply:
It's based on a timeout.  It's not once per session - but once you sudo once you get elevated privileges it sticks for a while (I think 30 minutes by default).  You can change it in the sudoers file to make it longer or shorter.